### PR TITLE
Delegate branch token serialization/deserialization to store layer

### PR DESCRIPTION
**Pull request by norberthu**
**What changed?**

**Why?**

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: common/persistence/persistenceMetricClients.go [hunk 1]
```diff
@@ -898,6 +898,36 @@ func (p *executionPersistenceClient) AppendRawHistoryNodes(
 	return resp, err
 }
 
+// ParseHistoryBranchInfo parses the history branch for branch information
+func (p *executionPersistenceClient) ParseHistoryBranchInfo(
+	ctx context.Context,
+	request *ParseHistoryBranchInfoRequest,
+) (*ParseHistoryBranchInfoResponse, error) {
+
+	branchInfo, err := ParseHistoryBranchToken(request.BranchToken)
```
Comment (yycptt) [inline common/persistence/persistenceMetricClients.go:907]: I don't follow. The call should be forward to `p.persistence`? Same Q for all three client impl.

File: common/persistence/dataInterfaces.go [hunk 2]
```diff
@@ -1196,6 +1223,26 @@ func UnixMilliseconds(t time.Time) int64 {
 	return unixNano / int64(time.Millisecond)
 }
 
+func ParseHistoryBranchToken(branchToken []byte) (*persistencespb.HistoryBranch, error) {
+	return serialization.HistoryBranchFromBlob(branchToken, enumspb.ENCODING_TYPE_PROTO3.String())
+}
```
Comment (yycptt) [inline common/persistence/dataInterfaces.go:1228]: Those two functions can be unified? I don't see any call to serialization.HistoryBranchFromBlob using an encoding different than PROTO3.

File: common/persistence/dataInterfaces.go [hunk 3]
```diff
@@ -1196,6 +1223,26 @@ func UnixMilliseconds(t time.Time) int64 {
...
+
+func UpdateHistoryBranchToken(branchToken []byte, branchInfo *persistencespb.HistoryBranch) ([]byte, error) {
+	bi, err := serialization.HistoryBranchFromBlob(branchToken, enumspb.ENCODING_TYPE_PROTO3.String())
```
Comment (yycptt) [inline common/persistence/dataInterfaces.go:1231]: For OSS impl, this step is unnecessary as all three defined fields are overwritten below?

File: common/persistence/history_manager.go [hunk 4]
```diff
@@ -127,13 +123,18 @@ func (m *executionManagerImpl) ForkHistoryBranch(
 		return nil, err
 	}
 
-	branchInfoBlob, err := m.serializer.HistoryBranchToBlob(newBranchInfo, enumspb.ENCODING_TYPE_PROTO3)
```
Comment (yycptt) [inline common/persistence/history_manager.go:130]: I probably lack some context on what needs to be customized here. But `m.serializer` is something that can be changed during initialization (search for `FactoryProvider` in persistence/factory/fx.go of your repo). Can that be used to achieve your goal?

File: common/persistence/persistenceMetricClients.go (hunk 1 above)
Comment (norberthu) [reply common/persistence/persistenceMetricClients.go:907]: darn... must have been a bizarre mishap (my mistake) when I rebased. This is not the code I intended! Let me fix this

File: common/persistence/dataInterfaces.go (hunk 3 above)
Comment (norberthu) [reply common/persistence/dataInterfaces.go:1231]: I'd argue this is much safer from a future-proof stand point. Branch tokens "may" evolve to include more stuff in the future. The intent here is to only overwrite the branch info portion of the branch token. Therefore we shouldn't assume branch info === branch token (logically speaking)

File: common/persistence/history_manager.go (hunk 4 above)
Comment (yycptt) [reply common/persistence/history_manager.go:130]: Synced offline. Currently here's no way to inject serializer for the one used in ndcHistoryReplicator, so have to do it this way for now. 

Once ^ is possible, then I think we should go with that route.

Comment (yycptt) [reply common/persistence/history_manager.go:130]: cc @yiminc What do you think?

**Review by yycptt — APPROVED** [review]

File: common/persistence/dataInterfaces.go (hunk 2 above)
Comment (yycptt) [reply common/persistence/dataInterfaces.go]: Is this resolved?

File: common/persistence/dataInterfaces.go (hunk 3 above)
Comment (yycptt) [reply common/persistence/dataInterfaces.go]: So the idea is that branchToken may contain some persistence store-specific information which may not in history branch definition? Even if that store-specific info is defined in OSS?

**Pull request merged**